src/business_logic/transaction/internal_objects.rs

```diff
@@ -166,3 +179,376 @@ impl InternalDeploy {
         todo!()
     }
 }
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+///  Represents an internal transaction in the StarkNet network that is a declaration of a Cairo
+///  contract class.
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+pub(crate) struct InternalDeclare {
+    pub(crate) class_hash: [u8; 32],
+    pub(crate) sender_address: Address,
+    pub(crate) tx_type: TransactionType,
+    pub(crate) validate_entry_point_selector: Felt,
+    pub(crate) version: u64,
+    pub(crate) max_fee: u64,
+    pub(crate) signature: Vec<Felt>,
+    pub(crate) nonce: Felt,
+    pub(crate) hash_value: Felt,
+}
+
+// ------------------------------------------------------------
+//                        Functions
+// ------------------------------------------------------------
+impl InternalDeclare {
+    pub fn new(
+        contract_class: ContractClass,
+        chain_id: Felt,
+        sender_address: Address,
+        max_fee: u64,
+        version: u64,
+        signature: Vec<Felt>,
+        nonce: Felt,
+    ) -> Result<Self, TransactionError> {
+        let hash = compute_class_hash(&contract_class)?;
+
+        let class_hash = hash
+            .to_bytes_be()
+            .try_into()
+            .map_err(|_| UtilsError::FeltToFixBytesArrayFail(hash.clone()))?;
+
+        let hash_value = calculate_declare_transaction_hash(
+            contract_class,
+            chain_id,
+            sender_address.clone(),
+            max_fee,
+            version,
+            nonce.clone(),
+        )?;
+
+        let validate_entry_point_selector = Felt::from_str_radix(
+            "1148189391774113786911959041662034419554430000171893651982484995704491697075",
+            10,
+        )
+        .unwrap();
+
+        let internal_declare = InternalDeclare {
+            class_hash,
+            sender_address,
+            tx_type: TransactionType::Declare,
+            validate_entry_point_selector,
+            version,
+            max_fee,
+            signature,
+            nonce,
+            hash_value,
+        };
+
+        internal_declare.verify_version()?;
+
+        Ok(internal_declare)
+    }
+
+    pub fn account_contract_address(&self) -> Address {
+        self.sender_address.clone()
+    }
+
+    pub fn validate_entrypoint_calldata(&self) -> Vec<Felt> {
+        let bytes = Felt::from_bytes_be(&self.class_hash);
+        Vec::from([bytes])
+    }
+
+    pub fn verify_version(&self) -> Result<(), TransactionError> {
+        // no need to check if its lesser than 0 because it is an usize
+        if self.version > u64::pow(2, 63) {
+            return Err(TransactionError::StarknetError(
+                "The sender_address field in Declare transactions of version 0, sender should be 1"
+                    .to_string(),
+            ));
```

Comment:
Same as with `FeeError` (here and the ones below).

Reply:
this error is very generic, i think it is not that bad to have it like this.